src/http/server/handlers/static_file_handler.cr

```diff
@@ -35,28 +35,30 @@ class HTTP::StaticFileHandler
     end
 
     original_path = context.request.path.not_nil!
-    is_dir_path = original_path.ends_with? "/"
-    request_path = self.request_path(URI.unescape(original_path))
 
+    unescaped_path = URI.unescape(original_path)
     # File path cannot contains '\0' (NUL) because all filesystem I know
     # don't accept '\0' character as file name.
-    if request_path.includes? '\0'
+    if unescaped_path.includes? '\0'
       context.response.status_code = 400
       return
     end
 
-    expanded_path = File.expand_path(request_path, "/")
-    if is_dir_path && !expanded_path.ends_with? "/"
-      expanded_path = "#{expanded_path}/"
+    request_path = self.request_path(unescaped_path)
+
+    is_dir_path = request_path.ends_with_separator?
+    expanded_path = request_path.expand("/")
```

Comment:
Why expand this here? Why not just join it with `public_dir` first then expand it?

Reply:
To make sure the path can't leave the `public_dir` path with a relative path starting with `../`. I don't know if this could be practically abused, but better safe than sorry.

Reply:
It does feel to me like there's too many lines of work here to do something "simple". Maintaining the trailing slash takes up a lot of this work. Is this a deficiency in the `Path` API?

Reply:
It might as well be. The issue is that an expanded path should be normalized and `normalize` scraps off trailing separators.
We could discuss if this is the expected behaviour for `normalize` and `expand`. Maybe `normalize` could have an option to preserve trailing slashes and use that in `expand`?

Reply:
Go [`path.Clean`](https://golang.org/pkg/path/#Clean) removes trailing slash. Node [`path.normalize`](https://nodejs.org/api/path.html#path_path_normalize_path) keeps it. Python's path class doesn't have a `normalize` method but it normalizes when expanding (using `os.path.normpath`) which keeps trailing slashes.
Java [org.apache.commons.io.FilenameUtils
](https://commons.apache.org/proper/commons-io/javadocs/api-2.4/org/apache/commons/io/FilenameUtils.html#normalize(java.lang.String)) interestingly has two different methods: `normalize` and `normalizeNoEndSeparator`.

Reply:
OK, let's keep trailing slashiness, but have an option to always remove it.